runtime/src/main/cpp/Exceptions.cpp

```diff
@@ -212,94 +212,127 @@ void ThrowException(KRef exception) {
+class {
+    /**
+     * Timeout 5 sec for concurrent (second) terminate attempt to give a chance the first one to finish.
+     * If the terminate handler hangs for 5 sec it is probably fatally broken, so let's do abnormal _Exit in that case.
+     */
+    unsigned int timeout = 5;
+    std::atomic_flag terminatingFlag = ATOMIC_FLAG_INIT;
+  public:
+    template <class Fun> RUNTIME_NORETURN void operator()(Fun block) {
+      if (!terminatingFlag.test_and_set()) {
+        block();
+        // block() is supposed to be NORETURN, otherwise go to normal abort()
+        konan::abort();
+      } else {
+        sleep(timeout);
+        // We come here when another terminate handler hangs for 5 sec, that looks fatally broken. Go to forced exit now.
+      }
+      _Exit(EXIT_FAILURE); // force exit
+    }
+} concurrentTerminateWrapper;
 
+void reportUnhandledException(KRef throwable) {
+  OnUnhandledException(throwable);
 #if KONAN_REPORT_BACKTRACE_TO_IOS_CRASH_LOG
-  {
-    LockGuard<SimpleMutex> lock(terminatingMutex);
-    if (!terminating) {
-      ReportBacktraceToIosCrashLog(throwable);
-    }
-  }
+  ReportBacktraceToIosCrashLog(throwable);
 #endif
+}
+
+}
 
-  konan::abort();
+RUNTIME_NORETURN void TerminateWithUnhandledException(KRef throwable) {
+  concurrentTerminateWrapper([=]() {
+    reportUnhandledException(throwable);
+    konan::abort();
+  });
 }
 
 // Some libstdc++-based targets has limited support for std::current_exception and other C++11 functions.
 // This restriction can be lifted later when toolchains will be updated.
 #if KONAN_HAS_CXX11_EXCEPTION_FUNCTIONS
 
-static void (*oldTerminateHandler)() = nullptr;
+namespace {
+class TerminateHandler {
+
+	// In fact, it's safe to call my_handler directly from outside: it will do the job and then invoke original handler,
+	// even if it has not been initialized yet. So one may want to make it public and/or not the class member
+  RUNTIME_NORETURN static void kotlinHandler () {
+    concurrentTerminateWrapper([]() {
+        if (auto currentException = std::current_exception()) {
+          try {
+            std::rethrow_exception(currentException);
+          } catch (ExceptionObjHolder &e) {
+            reportUnhandledException(e.obj());
+            konan::abort();
+          } catch (...) {
+            // Not a Kotlin exception - pass throw to default handler
+          }
+        }
+       // Come here in case of direct terminate() call or unknown exception - go to default terminate handler.
+ 			instance()->queuedHandler_();
+    });
+	}
 
-static void callOldTerminateHandler() {
-#if KONAN_REPORT_BACKTRACE_TO_IOS_CRASH_LOG
-  {
-    LockGuard<SimpleMutex> lock(terminatingMutex);
-    terminating = true;
-  }
-#endif
+  /// Use machinery like Meyers singleton to provide thread safety
 
-  RuntimeCheck(oldTerminateHandler != nullptr, "Underlying exception handler is not set.");
-  oldTerminateHandler();
-}
+  typedef __attribute__((noreturn)) void (*QH)();
```

Comment:
I would agree, but macro `RUNTIME_NORETURN` has its own issue: it hides the fact it is language extension, not the standard `[[NORETURN]]` as one may expect (or as it may change some day). However, in that particular case `[[NORETURN]]` will not work (function pointer), so I have to stay with gcc extension here, or use explicit cast. So at this point I want to may it clear (ie gcc extension vs. standard language feature). Moreover, I believe the macro `RUNTIME_NORETURN` is a sort of abuse: it reinvents language feature, hides important aspects and does not add value.